### PR TITLE
task 12 get/api/articles?topic endpoint with tests

### PR DESCRIPTION
**Pull request by AndrewLau97**
add topics query to /get/api/articles endpoint which filtered to only respond with articles of the given article
tested if it responds with filtered array of articles
no articles but topic exists
all three queries at the same time
topic query does not exist

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -221,30 +221,85 @@ describe("GET /api/articles", () => {
...
-      .get("/api/articles?sort_by=created")
-      .expect(400)
-      .then(({ body: { message } }) => {
-        expect(message).toBe("Bad request");
+    .get("/api/articles?topic=mitch")
+    .expect(200)
+    .then(({body:{articles}})=>{
+      expect(articles).toHaveLength(12);
+      articles.forEach((article)=>{
+        expect(article.topic).toBe("mitch")
```
Comment (nc-nathan-blanch) [inline __tests__/app.test.js:231]: nice 👍

File: __tests__/app.test.js [hunk 2]
```diff
@@ -221,30 +221,85 @@ describe("GET /api/articles", () => {
...
+      })
+    })
+  })
+  test("200: Responds with empty array when no articles of given topic",()=>{
```
Comment (nc-nathan-blanch) [inline __tests__/app.test.js:235]: nice test! what about when the topic doesn't exist?

File: __tests__/app.test.js [hunk 3]
```diff
@@ -221,30 +221,85 @@ describe("GET /api/articles", () => {
...
+    return request(app)
+    .get("/api/articles?topic=paper")
+    .expect(200)
+    .then(({body:{articles}})=>{
+      expect(articles).toHaveLength(0);
+    })
+  })
+  test("200: Takes in topic query, responds with all articles if topic is omitted",()=>{
```
Comment (nc-nathan-blanch) [inline __tests__/app.test.js:243]: I would say this is dealing with an edge case. If no topic is provided, you shouldn't need to circumnavigate it. Up to you though!

File: model/articles.models.js [hunk 4]
```diff
@@ -12,21 +12,36 @@ function fetchArticleById(article_id) {
...
+  const queryValues = [];
   let queryInsert = `SELECT articles.article_id, articles.author, title, topic, articles.created_at, articles.votes, article_img_url, COUNT(comments.body) AS comment_count 
  FROM articles LEFT JOIN comments
- ON articles.article_id=comments.article_id
- GROUP BY articles.article_id `;
-  queryInsert += `ORDER BY ${sort_by} ${order}`;
-  if(!validOrderQueries.test(order)||!validSortByQueries.includes(sort_by.toLowerCase())){
-    return Promise.reject({status:400,message:"Bad request"})
+ ON articles.article_id=comments.article_id `;
+  if (topic) {
```
Comment (nc-nathan-blanch) [inline model/articles.models.js:30]: good job :)

File: model/articles.models.js [hunk 5]
```diff
@@ -12,21 +12,36 @@ function fetchArticleById(article_id) {
...
+    queryInsert += `WHERE topic = $1 `;
+    queryValues.push(topic);
   }
-  return db.query(queryInsert).then(({rows}) => {
+  queryInsert += `GROUP BY articles.article_id 
+  ORDER BY ${sort_by} ${order}`;
+  if (
```
Comment (nc-nathan-blanch) [inline model/articles.models.js:36]: good error handling here, personally I prefer doing it at the very start of the function so that I don't accidentally query the database first!

**Review by nc-nathan-blanch — COMMENTED** [review]
This is good, I think you are missing one test though. See my comments. Also, try not to commit changes that aren't relevant to the ticket you're working on! They should be a separate branch + pull.
